glusterd2/transaction/context.go

```diff
@@ -110,23 +84,30 @@ func (c *Tctx) SetNodeResult(nodeID uuid.UUID, key string, value interface{}) er
 // Returns error if not found.
 func (c *Tctx) Get(key string, value interface{}) error {
 
-	storeKey := c.prefix + "/" + key
-	r, err := store.Store.Get(context.TODO(), storeKey)
-	if err != nil {
-		c.log.WithError(err).WithField("key", storeKey).Error("failed to get value from transaction context")
-		return err
+	// cache all keys and values from the store on the first call to Get
+	if !c.cached {
+		resp, err := store.Store.Get(context.TODO(), c.config.StorePrefix, clientv3.WithPrefix())
+		if err != nil {
+			c.logger.WithError(err).WithField("key", key).Error("failed to get key from transaction context")
+			return err
+		}
+		for _, kv := range resp.Kvs {
+			c.readSet[string(kv.Key)] = kv.Value
+		}
+		c.cached = true
 	}
 
-	if r.Count == 0 {
-		c.log.WithError(err).WithField("key", storeKey).Debug("key not found in store")
+	// return cached key
+	storeKey := c.config.StorePrefix + key
+	if data, ok := c.readSet[storeKey]; ok {
+		if err := json.Unmarshal(data, value); err != nil {
+			c.logger.WithError(err).WithField("key", storeKey).Error("failed to unmarshall value")
+		}
+	} else {
 		return errors.New("key not found")
```

Comment:
If you don't find a key in the cache you need to get it from the store. You cannot just return an error. The context might have had newer key-values set between the first Get and later Gets.

Reply:
> The context might have had newer key-values set between the first Get and later Gets.

I thought about that and it's very unlikely to be the case in real usage:

```go

txn.Set("somekey", something)

// do something 

txn.Get("somekey", &something)
```

The step functions run as a single isolated unit. The Get and Set in the previous code violated this. In the above snippet, there is no case for the `Get` to consume the key set by any other concurrently running step function. In effect, it'll only be expecting the value it set earlier. There's no reason do that as the value doesn't and shouldn't change.

Reply:
Refreshed the PR to aggregate Set() calls and defer committing to store. This should fix the problem you described.

Reply:
I did look at this a little later. You do always clear the cache each time a Set it done (though just locally), so the first Get later has to refresh the cache. My earlier comment becomes invalid.

But I don't agree with clearing out the full cache on a Set. That would throw away the gains to be had by caching. I'd suggest that you delete the cached key/value, and on a Get of the key, when you don't find it in the cache, get from store and cache it.